src/ros_robot_localization_listener.cpp

```diff
@@ -31,248 +31,359 @@
+
+  Eigen::AngleAxisd roll_angle(estimator_state.state(StateMemberRoll), Eigen::Vector3d::UnitX());
+  Eigen::AngleAxisd pitch_angle(estimator_state.state(StateMemberPitch), Eigen::Vector3d::UnitY());
+  Eigen::AngleAxisd yaw_angle(estimator_state.state(StateMemberYaw), Eigen::Vector3d::UnitZ());
+
+  Eigen::Quaterniond base_orientation(yaw_angle * pitch_angle * roll_angle);
+
+  Eigen::Affine3d base_pose(Eigen::Translation3d(base_position) * base_orientation);
+
+  // Now we can calculate the transform from odom to the requested frame (target)...
+  Eigen::Affine3d target_pose_odom = base_pose * target_pose_base;
+
+  // ... and put it in the output state
+  state(StateMemberX) = target_pose_odom.translation().x();
+  state(StateMemberY) = target_pose_odom.translation().y();
+  state(StateMemberZ) = target_pose_odom.translation().z();
+
+  Eigen::Vector3d ypr = target_pose_odom.rotation().eulerAngles(2, 1, 0);
+
+  state(StateMemberRoll)  = ypr[2];
+  state(StateMemberPitch) = ypr[1];
+  state(StateMemberYaw)   = ypr[0];
+
+  // Now let's calculate the twist of the target frame
+  // First get the base's twist
+  Twist base_velocity;
+  Twist target_velocity_base;
+  base_velocity.linear = Eigen::Vector3d(estimator_state.state(StateMemberVx),
+                                         estimator_state.state(StateMemberVy),
+                                         estimator_state.state(StateMemberVz));
+  base_velocity.angular = Eigen::Vector3d(estimator_state.state(StateMemberVroll),
+                                          estimator_state.state(StateMemberVpitch),
+                                          estimator_state.state(StateMemberVyaw));
+
+  // Then calculate the target frame's twist as a result of the base's twist.
+  /*
+   * We first calculate the coordinates of the velocity vectors (linear and angular) in the base frame. We have to keep
+   * in mind that a rotation of the base frame, together with the translational offset of the target frame from the base
+   * frame, induces a translational velocity of the target frame.
+   */
+  target_velocity_base.linear = base_velocity.linear + base_velocity.angular.cross(target_pose_base.translation());
+  target_velocity_base.angular = base_velocity.angular;
+
+  // Now we can transform that to the target frame
+  Twist target_velocity;
+  target_velocity.linear = target_pose_base.rotation().transpose() * target_velocity_base.linear;
+  target_velocity.angular = target_pose_base.rotation().transpose() * target_velocity_base.angular;
+
+  state(StateMemberVx) = target_velocity.linear(0);
+  state(StateMemberVy) = target_velocity.linear(1);
+  state(StateMemberVz) = target_velocity.linear(2);
+
+  state(StateMemberVroll) = target_velocity.angular(0);
+  state(StateMemberVpitch) = target_velocity.angular(1);
+  state(StateMemberVyaw) = target_velocity.angular(2);
+
+  // TODO: transform covariance
+  covariance = estimator_state.covariance;
```

Comment:
This covariance transformation is still a todo. A choice needs to be made though, because the rotational uncertainty in combination with translation implies a non-Gaussian uncertainty after transformation, so some approximation is needed. Do you have a suggestion, @ayrton04?